### PR TITLE
Port WarpCTC Operator

### LINKED ISSUE #4923: Port WarpCTC Operator.
- [WarpCTCLayer.cpp]( https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/gserver/layers/WarpCTCLayer.cpp)
- [WarpCTCLayer api](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/trainer_config_helpers/layers.py#L5257)

### PR DESCRIPTION
**Pull request by Xreki**
Fix #4923 
`check_grad` failed, so I disable it temporarily and will fix it next PR.

### CODE REVIEW

File: paddle/operators/warpctc_op.h [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+                      platform::dynload::ctcGetStatusString(status));
+    PADDLE_ENFORCE_GT(workspace_bytes, 0UL,
+                      "Bytes of workspace got by warp-ctc function, "
+                      "get_workspace_size(), should be larger than 0.");
+
+    Tensor workspace;
+    size_t workspace_elements = workspace_bytes / sizeof(float) + 1UL;
+    float* workspace_data = workspace.mutable_data<float>(
+        framework::make_ddim({static_cast<int64_t>(workspace_elements)}),
+        ctx.GetPlace());
```
Comment (qingqing01) [inline paddle/operators/warpctc_op.h]: Also can use `paddle::memory::Alloc` to allocate workspace.

File: paddle/operators/warpctc_op.h [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+                      "The number of sequences of Input(Logits) should be "
+                      "equal to that of Input(Label).");
+
+    const size_t sequence_width = logits->numel() / logits_dims[0];
+    auto loss_dims =
+        framework::make_ddim({static_cast<int64_t>(num_sequences), 1});
+
+    // warpctc needs sequences data stored in batch format
+    Tensor warpctc_logits;
+    math::PaddingSequenceFunctor<Place, T>()(ctx.device_context(), *logits,
```
Comment (qingqing01) [inline paddle/operators/warpctc_op.h]: im2col等functor里，output的内存分配是在functor外面，也就是op里面，在functor里调用`data<T>()`，而不是`mutable_data<T>()`延迟分配方式。

File: python/paddle/v2/framework/tests/test_warpctc_op.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        self.outputs = {"Loss": loss}
+        self.attrs = {"blank": blank, "normByTimes": norm_by_times}
+
+    def test_check_output(self):
+        self.check_output()
+
+
+#    def test_check_grad(self):
+#        self.outputs["WarpCTCGrad"] = None
+#        self.check_grad(["Logits"], "Loss", max_relative_error=0.01)
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_warpctc_op.py]: 为什么注释了梯度检测？

File: paddle/operators/warpctc_op.cc [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                   "Output(WarpCTCGrad) of WarpCTCOp should not be null.");
+    PADDLE_ENFORCE(ctx->HasOutput("Loss"),
+                   "Output(Loss) of WarpCTCOp should not be null.");
+
+    auto dims = ctx->GetInputDim("Logits");
+    int sequence_width = static_cast<int>(framework::product(dims) / dims[0]);
+    int blank = ctx->Attrs().Get<int>("blank");
+    PADDLE_ENFORCE((blank >= 0) && (blank < sequence_width),
+                   "The value of Attr(blank) should be in interval [0, %d).",
+                   sequence_width);
```
[the remark this answers is not in the document]
Comment (Xreki) [reply paddle/operators/warpctc_op.cc]: Because we cannot get the lod of inputs, so we cannot get the correct dimension of loss. I can set a wrong dimension here like `crf`, but it is meanless...

File: paddle/operators/warpctc_op.h (hunk 1 above)
Comment (Xreki) [reply paddle/operators/warpctc_op.h]: But I cannot use `SetConstant` to set the memory to 0. Is there any other interface to do this?

File: paddle/operators/warpctc_op.h (hunk 2 above)
Comment (Xreki) [reply paddle/operators/warpctc_op.h]: 这个主要是`max_sequence_length`一开始在`functor`里面算了。如果在外面分配好内存，`functor`里面还是需要检查`dimension`的啊，不过不一致就直接enforce挂掉？

File: python/paddle/v2/framework/tests/test_warpctc_op.py (hunk 3 above)
Comment (Xreki) [reply python/paddle/v2/framework/tests/test_warpctc_op.py]: 因为梯度检查不通过，我怀疑可能是`warp-ctc`只能作为最后一个op，不能接收来自上一个op的梯度的原因，还没具体查。下一个pr再查明吧。

File: paddle/operators/math/sequence_padding.cc [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+          T scale =
+              norm_by_times ? (1.0f / static_cast<T>(sequence_length)) : 1.0f;
+          for (size_t k = 0; k < sequence_width; ++k) {
+            padding_data[(i * num_sequences + j) * sequence_width + k] =
+                seq_data[(start_pos + i) * sequence_width + k] * scale;
+          }
+        } else {
+          memset(padding_data + (i * num_sequences + j) * sequence_width, 0,
+                 sequence_width * sizeof(T));
+        }
```
Comment (chengduoZH) [inline paddle/operators/math/sequence_padding.cc]: Line 63~66 should be removed, and setting `padding_data` to zero outside the loop.
```
math::SetConstant<platform::CPUPlace, T> set_zero;
set_zero(context, padding, static_cast<T>(0));
```

File: paddle/operators/math/sequence_padding.cu [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        sequence[sequence_base_idx + i] = scale * padding[padding_base_idx + i];
+      }
+    }
+  } else if (sequence_idx < max_sequence_length) {
+    if (Padding) {
+      /* sequence -> padding */
+      for (size_t i = threadIdx.x; i < sequence_width; i += blockDim.x) {
+        padding[padding_base_idx + i] = 0;
+      }
+    }
```
Comment (chengduoZH) [inline paddle/operators/math/sequence_padding.cu]: If following above [comment](https://github.com/PaddlePaddle/Paddle/pull/5107/files#diff-67db4d25265f59c2d923cd285180e62bR66), lines of 50~56 can be removed.

File: paddle/operators/math/sequence_padding.cc (hunk 5 above)
Comment (Xreki) [reply paddle/operators/math/sequence_padding.cc]: I am not sure. A previous `set_zero` will double the memory access and may increase the time cost.

File: paddle/operators/math/sequence_padding.cc [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+limitations under the License. */
+
+#include "paddle/operators/math/sequence_padding.h"
+
+namespace paddle {
+namespace operators {
+namespace math {
+
+template <typename T>
+class PaddingSequenceFunctor<platform::CPUPlace, T> {
```
[the remark this answers is not in the document]
Comment (Xreki) [reply paddle/operators/math/sequence_padding.cc]: Done.

File: paddle/operators/math/sequence_padding.cc [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+ public:
+  void operator()(const platform::DeviceContext& context,
+                  const framework::LoDTensor& seq, framework::Tensor& padding,
+                  bool norm_by_times) {
+    auto lod = seq.lod();
+    PADDLE_ENFORCE_GT(lod.size(), 0UL,
+                      "The lod of LoDTensor seq should not be null.");
```
[the remark this answers is not in the document]
Comment (Xreki) [reply paddle/operators/math/sequence_padding.cc]: Done.

File: paddle/operators/math/sequence_padding.h [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+ *        seq[lod[level][i]] = padding[i]
+ *
+ *  All sequences will be padded to the same length.
+ *  Example:
+ *    seq     (s0, s0, s0, s0; s1, s1; s2, s2, s2; s3)
+ *    padding (s0, s1, s2, s3; s0, s1, s2, 0; s0, 0, s2, 0; s0, 0, 0, 0)
+ *
+ * \param context       device context of this functor.
+ * \param seq           LoDTensor which is stored in sequence format.
+ * \param padding       Tensor which is padded to the same length.
```
[the remark this answers is not in the document]
Comment (Xreki) [reply paddle/operators/math/sequence_padding.h]: Done.

File: paddle/operators/warpctc_op.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (Xreki) [reply paddle/operators/warpctc_op.cc]: Like crf, I set a wrong dimension here and correct the dimension in the kernel.

File: paddle/operators/warpctc_op.cc [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  framework::DataType IndicateDataType(
```
[the remark this answers is not in the document]
Comment (Xreki) [reply paddle/operators/warpctc_op.cc]: Updated.

File: paddle/operators/warpctc_op.cc [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+class WarpCTCOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  WarpCTCOpMaker(framework::OpProto* proto,
+                 framework::OpAttrChecker* op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("Logits",
+             "(LodTensor, default: LoDTensor<float>), the unscaled "
+             "probabilities of variable-length sequences, which is a 2-D "
+             "Tensor with LoD information. It is of the shape "
```
[the remark this answers is not in the document]
Comment (Xreki) [reply paddle/operators/warpctc_op.cc]: Done.

File: paddle/operators/warpctc_op.h (hunk 2 above)
Comment (Xreki) [reply paddle/operators/warpctc_op.h]: Done. 改成在Functor外面分配好空间，Functor里面只做check了。

**Review by qingqing01 — DISMISSED** [review]
Approved this PR. Before merging this PR, please create an issue about the gradient check.

**Review by qingqing01 — APPROVED** [review]
